cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java

```diff
@@ -134,10 +137,15 @@ public Counter addRequestAndGetCount(ProgramRunId programRunId) throws Exception
    *
    * @param programRunId run id associated with the launch request
    */
-  public void addRequest(ProgramRunId programRunId) {
-    launchingQueue.add(programRunId);
+  public int addRequest(ProgramRunId programRunId) {
+    int result;
+    synchronized (this) {
```

Comment:
nit: I'd synchronize on the queue itself. It's usually safer as you never know who may want to synchronize on yourself